UefiTestingPkg/PerfTests/ServicesBenchmarkTest/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+//! Services Benchmark Test Library
+//!
+//! This crate provides a set of benchmarks for measuring the performance of various UEFI services.
+//! It is intended to be run in a UEFI environment to collect timing and call statistics for selected
+//! UEFI service functions. The results are output in a markdown-formatted table for easy analysis.
+//!
+//! ## Usage
+//!
+//! Invoke the `bench_start` function from your UEFI application or test harness, passing the UEFI
+//! image handle and system table. The library will execute a set of predefined benchmarks and print
+//! the results to the UEFI console.
+//!
+//! ## Output
+//! 
+//! The benchmark results include the name of each tested service, total cycles consumed, number of calls,
+//! and average cycles per operation.
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation. All rights reserved.
+//!
+//! SPDX-License-Identifier: BSD-2-Clause-Patent
+#![cfg_attr(target_os = "uefi", no_std)]
+
+extern crate alloc;
+use alloc::vec::Vec;
+use core::fmt::Write;
+use mu_rust_helpers::perf_timer::{Arch, ArchFunctionality as _};
+
+use patina::boot_services::StandardBootServices;
+use r_efi::{efi, system};
+use rust_advanced_logger_dxe::{DEBUG_ERROR, DEBUG_INFO, debugln};
+
+use crate::{error::BenchError, measure::BENCH_FNS};
+use alloc::string::String;
+
+/// Global instance of UEFI Boot Services.
+pub static BOOT_SERVICES: StandardBootServices = StandardBootServices::new_uninit();
+
+pub fn bench_start(handle: efi::Handle, st: *const system::SystemTable) -> Result<(), BenchError> {
+    debugln!(DEBUG_INFO, "Starting Services Benchmark Test...");
+
+    let mut output_buf = String::new();
+
+    // Writes fixed-width markdown table.
+    // Column headers.
+    writeln!(
+        &mut output_buf,
+        "| {:<32} | {:>14} | {:>12} | {:>15} | {:>15} | {:>12} | {:>12} | {:>12} |",
+        "Name",
+        "Total cycles",
+        "Total calls",
+        "Cycles/op",
+        "Total time (ms)",
+        "Min cycles",
+        "Max cycles",
+        "SD [cycles]"
+    )
+    .map_err(|e| BenchError::WriteFailure("Write table header failed", e))?;
+    // Column seperators.
+    writeln!(
+        &mut output_buf,
+        "| {:-<32} | {:-<14} | {:-<12} | {:-<15} | {:-<15} | {:-<12} | {:-<12} | {:-<12} |",
+        "-", "-", "-", "-", "-", "-", "-", "-"
+    )
+    .map_err(|e| BenchError::WriteFailure("Write table header failed", e))?;
+
+    for (bf, num_calls) in BENCH_FNS {

```

Comment:
I'd split this up into a separate function and gather the results into a vector of result structs or something, then report the results